crates/hir_ty/src/infer/coerce.rs

```diff
@@ -324,12 +328,45 @@ impl<'a> InferenceContext<'a> {
                 return Err(err);
             }
         };
-        // FIXME: record overloaded deref adjustments
-        success(
-            vec![Adjustment { kind: Adjust::Borrow(AutoBorrow::Ref(to_mt)), target: ty.clone() }],
-            ty,
-            goals,
-        )
+        if ty == from_ty && from_mt == Mutability::Not && autoderef.step_count() == 1 {
+            // As a special case, if we would produce `&'a *x`, that's
+            // a total no-op. We end up with the type `&'a T` just as
+            // we started with.  In that case, just skip it
+            // altogether. This is just an optimization.
+            //
+            // Note that for `&mut`, we DO want to reborrow --
+            // otherwise, this would be a move, which might be an
+            // error. For example `foo(self.x)` where `self` and
+            // `self.x` both have `&mut `type would be a move of
+            // `self.x`, but we auto-coerce it to `foo(&mut *self.x)`,
+            // which is a borrow.
+            always!(to_mt == Mutability::Not); // can only coerce &T -> &U
+            return success(vec![], ty, goals);
+        }
+
+        let mut adjustments = self.auto_deref_adjust_steps(&autoderef);
+        adjustments
+            .push(Adjustment { kind: Adjust::Borrow(AutoBorrow::Ref(to_mt)), target: ty.clone() });
+
+        success(adjustments, ty, goals)
+    }
+
+    pub(super) fn auto_deref_adjust_steps(&self, autoderef: &Autoderef<'_>) -> Vec<Adjustment> {
+        let steps = autoderef.steps();
+        let targets =
+            steps.iter().skip(1).map(|(_, ty)| ty.clone()).chain(iter::once(autoderef.final_ty()));
+        steps
+            .iter()
+            .map(|(kind, _source)| match kind {
+                AutoderefKind::Overloaded => {
+                    // FIXME: mutability
+                    Some(OverloadedDeref(Mutability::Not))
```

Comment:
any idea how to go about this?

Reply:
I'm a bit confused about what rustc does. [This](https://github.com/rust-lang/rust/blob/a84d1b21aea9863f0fc5f436b4982d145dade646/compiler/rustc_typeck/src/check/autoderef.rs#L54-L57) seems to be the equivalent place unless I'm missing something, which resolves the overloaded deref and then uses the mutability of the return value; but that just calls [this](https://github.com/rust-lang/rust/blob/a84d1b21aea9863f0fc5f436b4982d145dade646/compiler/rustc_typeck/src/check/place_op.rs#L139-L143), which says it only deals with `Deref`. So I would conclude that this is never mutable. But maybe I'm missing something. It seems the derefs only get converted to mutable in a [separate step](https://github.com/rust-lang/rust/blob/a84d1b21aea9863f0fc5f436b4982d145dade646/compiler/rustc_typeck/src/check/place_op.rs#L221-L228), which gets called depending on the place (i.e. if we have `foo.bar = baz;`, the autoderefs for `foo` would get converted). We don't track whether we need a mutable place during inference yet either, that's also something we should start at some point :smile:

Reply:
Ye I saw the same stuff there and was kinda confused hence my asking, Makes sense that this is changed up in a different step.